_posts/2023-07-14-types-and-programming-languages.markdown

```diff
@@ -0,0 +1,610 @@
+If you are looking for some advanced concepts, feel free to continue reading! Beware, math ahead!
+
+Sometimes, modeling the evaluator and the type-checker as functions is not enough. One wants to model them as relations, and determine some properties about these relations, such as the order of evaluation being irrelevant for the final result.
+
+In the rest of this blog post, largely inspired by the book "Types and Programming Languages", Chapter 8, written by Benjamin Pierce, I will illustrate one element of the proof: the one that inductive and constructive versions of the set of Terms are equivalent. Having equivalence enables obtaining other results out of the scope of this blog post, including that the order of evaluation does not matter.
+
+With the help of this trick, it becomes possible to prove similar equivalences for different inductive and constructive definitions of:
+- The set of `(Expr, Expr)` of small-step evaluations
+- The set of `(Expr, Type)` of type checking
+
+but I leave these as an exercise for the interested reader.
+
+In Types and Programming Languages, chapter 3.2, we discover that there are two other mathematical definitions of the "set of all terms".
+The first one in definition 3.2.1 states that the set of _terms_ is the smallest set 𝒯 such that:
+
+1. $$\{\texttt{true}, \texttt{false}, 0\} \subseteq 𝒯$$;
+2. if $$t_1 \in 𝒯$$, then $$\{\texttt{succ}\;t_1, \texttt{pred}\;t_1, \texttt{is_zero}\;t_1\} \subseteq 𝒯$$;
+3. if $$t_1 \in 𝒯$$, $$\;\; t_2 \in 𝒯$$ and $$t_3 \in 𝒯$$, then $$\texttt{if}\;t_1\;\texttt{then}\;t_2\;\texttt{else}\;t_3 \in 𝒯$$.
+
+Note that these terms omit `Double` and `Add` above. This means we cannot state that this set is the same as `set t: Term | true` as one would like to write, but let's continue.
+
+We can write the inductive definition above in Dafny too:
+
+{% highlight javascript %}
+ghost const AllTermsInductively: iset<Term>
+
+ghost predicate InductionCriteria(terms: iset<Term>) {
+  && True in terms && False in terms && Zero in terms
+  && (forall t1 <- terms :: Succ(t1) in terms)
+  && (forall t1 <- terms :: Pred(t1) in terms)
+  && (forall t1 <- terms :: IsZero(t1) in terms)
+  && (forall t1 <- terms, t2 <- terms, t3 <- terms
+        :: If(t1, t2, t3) in terms)
+}
+lemma {:axiom} InductiveAxioms()
+  ensures InductionCriteria(AllTermsInductively)
+  ensures forall setOfTerms: iset<Term> | InductionCriteria(setOfTerms)
+            :: AllTermsInductively <= setOfTerms
+{% endhighlight %}
+
+The second definition for the set of all terms in section 3.2.3 is done constructively. We first define a set $$S_i$$ for each natural number $$i$$, as follows
+
+$$S_0 = ∅ $$;  
+$$\begin{aligned}S_{i+1} = && && \{\texttt{true}, \texttt{false}, 0\} \\ && \bigcup && \{\texttt{succ}\;t_1, \texttt{pred}\;t_1, \texttt{is_zero}\;t_1 \mid t_1 \in S_i \} \\ && \bigcup && \{\texttt{if}\;t_1\;\texttt{then}\;t_2\;\texttt{else}\;t_3\mid t_1 \in S_i,\; t_2 \in S_i,\; t_3 \in S_i\}\end{aligned}$$.
```

Comment:
In a few para below, "The second definition for the set of all terms in section 3.2.3...", there is a ' . ' after the 2nd union, but not anywhere else. 
I do see you already have had a discussion about it.